### PR TITLE
feat: http + json extended example

### LINKED ISSUE #8: example(json): disclose values out of `spotify_response.json`

### PR DESCRIPTION
**Pull request by lonerapier**
Closes #8 

Adds:
- codegen for http + json circuit
- build docs for extended example
- http docs
- refactoring codegen to make it more modular for folded circuits
- test harness
- add test and lint CI

Run `npx mocha -g spotify` to run spotify example. Currently, we don't match the output because `sym` file is too big to be loaded, due to unconstrained signals. However, running circuit manually with circom, snarkjs works.

To run codegen, which will generate circuit in `main/spotify_top_artists.circom`: 
```
pabuild codegen extended --circuit-name spotify_top_artists --input-file examples/http/spotify_top_artists.json --lockfile examples/http/lockfile/spotify_extended.lock.json -d
```

and then compile circuits and generate witness using circomkit.

### CODE REVIEW

File: examples/http/spotify_top_artists_request.http [hunk 1]
```diff
@@ -0,0 +1,3 @@
+GET /v1/me/top/artists?time_range=medium_term&limit=1 HTTP/1.1
+Host: api.spotify.com
+Authorization: Bearer BQBXRpIm2NL08akEiaxB5l42eiq6Zd9Q0S2-0Q4k0CMoa5u8o_ah_Ddjxt6Mv3226AEDyKYcFPpgw_6Asg-Y2hJpcuMya8wzqyqgiV-KH0vcEq7EFzODXoaBxsB0wryVCWDF6p5dqcpIHOz4QJqQa9mUA6sFzYNyECglT-BGcRe_N9f_3aqYTGQ-kkE-devPkPkEfDcbziT6mOzJfGRzLw
```
Comment (devloper) [inline examples/http/spotify_top_artists_request.http:3]: Do we want to commit this token...?

File: examples/http/lockfile/spotify.lock.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+    "version": "HTTP/1.1",
+    "status": "200",
+    "message": "OK",
+    "headerName1": "content-type",
+    "headerValue1": "application/json; charset=utf-8",
+    "headerName2": "content-encoding",
+    "headerValue2": "gzip"
```
Comment (devloper) [inline examples/http/lockfile/spotify.lock.json:8]: This is interesting. If the content is gzip vs identity, this likely fails.

**Review by devloper — COMMENTED** [review]
few thoughts

File: examples/http/lockfile/spotify.lock.json (hunk 2 above)
Comment (lonerapier) [reply examples/http/lockfile/spotify.lock.json:8]: we can use `Accept-Encoding: identity` in request and perform `LockHeader` proof on request, right?

File: examples/http/spotify_top_artists_request.http (hunk 1 above)
Comment (lonerapier) [reply examples/http/spotify_top_artists_request.http:3]: It expires every hour, at this point even I can't use it lol. but, yeah, will add this to gh secrets.

**Review by Autoparallel — APPROVED** [review]
looks good to me! Lot's of good additions and clean up.

I think my review may be a bit superficial as I want to get to use this outside of this context, that will really be the most telling if we have this working in the best way possible :)

**Comment by Autoparallel** [comment]
Also note I pushed up a README change.

**Comment by Autoparallel** [comment]
@lonerapier see #92 to see if that fixes test

**Pull request merged**